src/client/chat.js

```diff
@@ -22,6 +24,21 @@ module.exports = function (client, options) {
   client._lastChatSignature = null
   client._lastRejectedMessage = null
 
+  // 1.20.3+ serializes chat components in chat packets with NBT. Non-chat packets that send messages (like disconnect) still use JSON chat.
+  // NMP API expects a JSON string, so since the schema is mostly the same we can convert the NBT to JSON with a transform to UUID encoding
+  function handleNbtComponent (nbtDataOrString) {
```

Comment:
I think we should expose that function here or in pchat so that mineflayer and flying-squid can use it

Reply:
I added prismarine-chat to the test code so it's cleaner (as opposed to handling chat internally), but otherwise there isn't much to add to prismarine-chat. The `handleNbtComponent` basically just calls simplify and does some mapping for UUIDs.

But I guess there could be a `ChatMessage.fromNbtComponent()` inside prismarine-chat to get a ChatMessage directly. But would it would require a change in nmp API as `playerChat` and `systemChat` current expect JSON strings

Reply:
my concern is how do we handle the "chat nbt" in the 10 other non-chat packets where that "chat nbt" is used in mineflayer?

I can see these solutions
1. we duplicate handleNbtComponent in nmp and mineflayer and flying-squid (and any other nmp users)
2. we expose handleNbtComponent in nmp and use them in users
3. we expose handleNbtComponent in pchat and use it in nmp, mineflayer, flying-squid and other places
4. we expose handleNbtComponent in pnbt and use it in all users


I think 3 or 4 are better.
2 is okay

1 I don't think it's a good idea

Reply:
https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.20.3/protocol.json look for all occurences of anonymousNbt

Reply:
Ah yeah I see, looks like will need some changes to all node-minecraft-protocol, prismarine-chat, mineflayer to swap the event from JSON string to a ChatMessage for #4. Maybe updating fromNotch() to and then emitting the ChatMessage:

```js
    static fromNotch (msg) {
      if (registry.supportFeature('chatPacketsUseNbtComponents') && msg.type) {
        const simplified = nbt.simplify(msg)
        const json = JSON.stringify(simplified, (key, val) => {
          if (key === 'id' && Array.isArray(val)) return uuidFromIntArray(val)
          return val
        })
        return new ChatMessage(JSON.parse(json))
      } else try {
        return new ChatMessage(JSON.parse(msg))
      } catch (e) {
        return new ChatMessage(msg)
      }      
    }
```

However doing this would break the fromNetwork in prismarine-chat as it expects a JSON object when formatting the chat string, so that too would need to be updated

Reply:
maybe we should expose an independent method in pchat

Reply:
but yeah hmm let's figure out how to unblock this